Sources/TSCUtility/PolymorphicCodable.swift

```diff
@@ -0,0 +1,56 @@
+/*
+ This source file is part of the Swift.org open source project
+
+ Copyright (c) 2014 - 2020 Apple Inc. and the Swift project authors
+ Licensed under Apache License v2.0 with Runtime Library Exception
+
+ See http://swift.org/LICENSE.txt for license information
+ See http://swift.org/CONTRIBUTORS.txt for Swift project authors
+*/
+
+/// Allows encoding and decoding known polymorphic types.
+public protocol PolymorphicCodableProtocol: Codable {
+    static var implementations: [PolymorphicCodableProtocol.Type] { get }
+}
+
+@propertyWrapper
+public struct PolymorphicCodable<T: PolymorphicCodableProtocol>: Codable {
+    public let value: T
+
+    public init(wrappedValue value: T) {
+        self.value = value
+    }
+
+    public var wrappedValue: T {
+        return value
+    }
+
+    public func encode(to encoder: Encoder) throws {
+        var container = encoder.unkeyedContainer()
+        try container.encode(String(reflecting: type(of: value)))
+        try container.encode(value)
+    }
+
+    public init(from decoder: Decoder) throws {
+        var container = try decoder.unkeyedContainer()
+        let typeCode = try container.decode(String.self)
+        guard let klass = T.implementations.first(where: { String(reflecting: $0) == typeCode }) else {
```

Comment:
Should this error or at least assert if there are duplications, to detect programmer error?

Reply:
if implementations was a Set then we (eventually) could `typeHint` -> `_typeByName` to then `T.allowedImplementations.contains(thatTypeByName)` to avoid that scanning through. The T we need anyway in order to get the `Any.Type as Codable` to call decoding on it

Reply:
I'm exploring another direction for registering types on Joe's hints... I'll report back once I have something.

More based around "intent" or "permissions" i.e. "this type is intended for crossing network" vs. "this one isnt" etc.